geode-wan-txgrouping/src/main/java/org/apache/geode/cache/wan/internal/txgrouping/serial/TxGroupingSerialGatewaySenderQueue.java

```diff
@@ -126,4 +148,110 @@ protected void postProcessBatch(final List<AsyncEvent<?, ?>> batch, final long l
         lastKey);
   }
 
+  /**
+   * This method returns a list of objects that fulfill the matchingPredicate
+   * If a matching object also fulfills the endPredicate then the method
+   * stops looking for more matching objects.
+   */
+  protected List<KeyAndEventPair> getElementsMatching(Predicate<GatewaySenderEventImpl> condition,
+      Predicate<GatewaySenderEventImpl> stopCondition,
+      long lastKey) {
+    GatewaySenderEventImpl event;
+    List<KeyAndEventPair> elementsMatching = new ArrayList<>();
+
+    long currentKey = lastKey;
+
+    while ((currentKey = inc(currentKey)) != getTailKey()) {
+      if (extraPeekedIds.contains(currentKey)) {
+        continue;
+      }
+      event = (GatewaySenderEventImpl) optimalGet(currentKey);
+      if (event == null) {
+        continue;
+      }
+
+      if (condition.test(event)) {
+        elementsMatching.add(new KeyAndEventPair(currentKey, event));
+
+        if (stopCondition.test(event)) {
+          break;
+        }
+      }
+    }
+
+    return elementsMatching;
+  }
+
+  @Override
+  public synchronized void remove() throws CacheException {
```

Comment:
At first glance it looks like a lot of the remove method content is reimplemented here. Is there a way break up the original with some extension points to deduce the duplication?

Reply:
See https://github.com/Nordix/geode/pull/17

Reply:
Nice!